### PR TITLE
Additional vector functions

### PR DESCRIPTION
**Pull request by gilesknap**
I have found it useful to add vector and scalar operator overrides for + - * and /

so you can do this for example:
```python
class Anchor3(Enum):
    """Anchor point for cuboids."""

    MIDDLE = 'middle'
    BOTTOM_MIDDLE = 'bottom_middle'
    TOP_MIDDLE = 'top_middle'

    BOTTOM_SW = 'bottom_sw'
    BOTTOM_NW = 'bottom_nw'
    BOTTOM_NE = 'bottom_ne'
    BOTTOM_SE = 'bottom_se'

    TOP_SW = 'top_sw'
    TOP_NW = 'top_nw'
    TOP_NE = 'top_ne'
    TOP_SE = 'top_se'


class Volume:
    """
    Describes a 3d space in a Minecraft world using a starting point and
    size. The starting point can be any vertex or the middle of one of the
    horizontal faces, using the cardinal terminology defined in Anchor3
    """
    def __init__(self, start: Vec3, size: Vec3, anchor: Anchor3 = Anchor3.BOTTOM_SW):
        if anchor == Anchor3.BOTTOM_SW:
            # start to start+size from bottom SW already describes the volume
            pass
        elif anchor == Anchor3.BOTTOM_NW:
            start += Vec3(0, 0,  + size.z - 1)
        elif anchor == Anchor3.BOTTOM_MIDDLE:
            start -= Vec3(1, 0, 1) * (size / 2).to_int()
        else:
            # TODO support remaining Anchor3
            raise ValueError("unsupported anchor")

        # invert z so posiitve size is positive North, decrement to include start block
        self.end = start + (size - 1) * Vec3(1, 1, -1)
        self.start = start
        self.size = size

    def fill(self, client: Client, block: Item = Item.AIR):
        if self.size.volume < 32768:
            client.fill(self.start, self.end, block.value)
        else:
            profile = [[block.value] * int(self.size.x)] * int(self.size.y)
            mktunnel(
                client,
                profile,
                self.start,
                direction=Direction.UP,
                anchor=Anchor.BOTTOM_LEFT,
                length=int(self.size.dz)
            )

```

### CODE REVIEW

**Comment by gilesknap** [comment]
The new CI is failing but I cannot follow the link to details, cant route to  http://172.23.56.1:8090/job/mcwb/job/PR-3/1/display/redirect

**Comment by conqp** [comment]
> The new CI is failing but I cannot follow the link to details, cant route to http://172.23.56.1:8090/job/mcwb/job/PR-3/1/display/redirect

Yeah, I just yesterday migrated to a private Jenkins + SonarQube setup. I'm still figuring things out, but I got the checks running now.

File: mcwb/types.py [hunk 1]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
     y: Number = 0
     z: Number = 0
 
-    def __add__(self, other: Vec3):
-        return type(self)(self.x + other.x, self.y + other.y, self.z + other.z)
+    def __add__(self, other):
+        if type(other) is Vec3:
```
Comment (conqp) [inline mcwb/types.py:44]: Please use `isinstance()` check instead.

File: mcwb/types.py [hunk 2]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x + other.x, self.y + other.y, self.z + other.z)
+        else:
```
Comment (conqp) [inline mcwb/types.py:46]: Please remove unneccessary else after return.

File: mcwb/types.py [hunk 3]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x + other, self.y + other, self.z + other)
+
+    def __sub__(self, other):
+        if type(other) is Vec3:
```
Comment (conqp) [inline mcwb/types.py:50]: Please use `isinstance()` check instead.

File: mcwb/types.py [hunk 4]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x - other.x, self.y - other.y, self.z - other.z)
+        else:
```
Comment (conqp) [inline mcwb/types.py:52]: Please remove unneccessary else after return.

File: mcwb/types.py [hunk 5]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x - other, self.y - other, self.z - other)
+
+    def __mul__(self, other):
+        if type(other) is Vec3:
```
Comment (conqp) [inline mcwb/types.py:56]: Please use `isinstance()` check instead.

File: mcwb/types.py [hunk 6]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x * other.x, self.y * other.y, self.z * other.z)
+        else:
```
Comment (conqp) [inline mcwb/types.py:58]: Please remove unneccessary else after return.

File: mcwb/types.py [hunk 7]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x * other, self.y * other, self.z * other)
+
+    def __truediv__(self, other):
+        if type(other) is Vec3:
```
Comment (conqp) [inline mcwb/types.py:62]: Please use `isinstance()` check instead.

File: mcwb/types.py [hunk 8]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x / other.x, self.y / other.y, self.z / other.z)
+        else:
```
Comment (conqp) [inline mcwb/types.py:64]: Please remove unneccessary else after return.

File: mcwb/types.py [hunk 9]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+            return type(self)(self.x / other, self.y / other, self.z / other)
+
+    def to_int(self):
```
Comment (conqp) [inline mcwb/types.py:67]: I am not too fond of the name `to_int` as it does not return an `int`, but a `Vec3` with ints.
Maybe `with_ints()` or `coerce_int()` would be better.
Also please add a docstring.

File: mcwb/types.py [hunk 10]
```diff
@@ -40,14 +40,36 @@ class Vec3(NamedTuple):
...
+        return type(self)(int(self.x), int(self.y), int(self.z))
 
-    def __sub__(self, other: Vec3):
-        return type(self)(self.x - other.x, self.y - other.y, self.z - other.z)
-
-    def __mul__(self, other: Number):
-        return type(self)(self.x * other, self.y * other, self.z * other)
+    @property
+    def volume(self):
```
Comment (conqp) [inline mcwb/types.py:71]: Please add a docstring.

**Review by conqp — CHANGES_REQUESTED** [review]

**Comment by gilesknap** [comment]
Fair comments - I believe 5ba8def resolves them

**Comment by gilesknap** [comment]
The CI looks good! do you want me to add some tests for the new vector functions?

**Review by conqp — APPROVED** [review]
Thanks for the fixes. Looks good.

**Comment by conqp** [comment]
> The CI looks good! do you want me to add some tests for the new vector functions?

This would be greatly appreciated. I am currently still busy writing unit tests for `mcipc`. If you want to implement tests for `mcwb`, you're more than welcome to do so.

**Pull request merged**